framework/src/main/java/org/checkerframework/framework/util/SimpleQualifierHierarchy.java

```diff
@@ -0,0 +1,287 @@
+                        kind);
+            }
+            quals.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+        }
+        return Collections.unmodifiableMap(quals);
+    }
+
+    protected Set<AnnotationMirror> createTops() {
+        Set<AnnotationMirror> tops = AnnotationUtils.createAnnotationSet();
+        for (Map.Entry<QualifierKind, AnnotationMirror> entry : qualifierMap.entrySet()) {
+            if (entry.getKey().isTop()) {
+                tops.add(entry.getValue());
+            }
+        }
+        return Collections.unmodifiableSet(tops);
+    }
+
+    protected Set<AnnotationMirror> createBottoms() {
+        Set<AnnotationMirror> bottoms = AnnotationUtils.createAnnotationSet();
+        for (Map.Entry<QualifierKind, AnnotationMirror> entry : qualifierMap.entrySet()) {
+            if (entry.getKey().isBottom()) {
+                bottoms.add(entry.getValue());
+            }
+        }
+        return Collections.unmodifiableSet(bottoms);
+    }
+
+    protected QualifierKind getQualifierKind(AnnotationMirror anno) {
+        String name = AnnotationUtils.annotationName(anno);
+        QualifierKind kind = qualifierKindHierarchy.getQualifierKind(name);
+        if (name == null) {
+            throw new BugInCF("Annotation not in hierarchy: %s", anno);
+        }
+        return kind;
+    }
+
+    @Override
+    public AnnotationMirror findAnnotationInHierarchy(
+            Collection<? extends AnnotationMirror> annos, AnnotationMirror top) {
+        return findAnnotationInSameHierarchy(annos, top);
+    }
+
+    @Override
+    public AnnotationMirror findAnnotationInSameHierarchy(
+            Collection<? extends AnnotationMirror> annos, AnnotationMirror annotationMirror) {
+        QualifierKind kind = getQualifierKind(annotationMirror);
+        for (AnnotationMirror anno : annos) {
+            QualifierKind annoKind = getQualifierKind(anno);
+            if (annoKind.areInSameHierarchy(kind)) {
+                return anno;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getTopAnnotations() {
+        return tops;
+    }
+
+    @Override
+    public AnnotationMirror getTopAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return qualifierMap.get(kind.getTop());
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getBottomAnnotations() {
+        return bottoms;
+    }
+
+    @Override
+    public AnnotationMirror getBottomAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return qualifierMap.get(kind.getBottom());
+    }
+
+    @Override
+    public AnnotationMirror getPolymorphicAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        QualifierKind poly = qualifierKindHierarchy.getTopToPoly().get(kind);
```

Comment:
This code assumes that `kind` is a key in `getTopToPoly()`, but there is no guarantee that `start` is a top.  (The specification of this method in `QualifierHierarchy` does not demand it.)